### PR TITLE
Api guidelines

### PR DESCRIPTION
**Pull request by barbara79**
This is a first part of API Guidelines project any suggestion is well accepted ;)

Please feel free to add other members of your team

https://github.com/commercetools/commercetools-api-reference/issues/152

### CODE REVIEW

File: API-GUIDELINE.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+A resource may contain sub-collection resources.
+
+##Methods
+The HTTP verbs used are;
+
+ - GET -> retrieve data
+ - POST -> create or update data
+ - DELETE -> delete data
+
+It's important to keep in mind the concept of Idempotency. So for instance use GET to retrieve data and not to delete content!
```
Comment (jenschude) [inline API-GUIDELINE.md:51]:
```suggestion
It's important to avoid unexpected behavior. So for instance use GET to retrieve data and not to delete content!
```

Idempotency is something different and mostly affects only update operations. Most of our updates are not idempotent. Cause either they produce errors or sending twice the same update doesn't result in the same response.

File: API-GUIDELINE.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+Define a Number/Integer property using names that can be quantified.
+
+Some examples:
+
+        availableQuantity
+        restockableInDays
+        version
+        weight
+        height
+        orderNumber
```
Comment (jenschude) [inline API-GUIDELINE.md:213]: orderNumber is a string in our API ;)

File: API-GUIDELINE.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+Some examples:
+
+      orderEditResult
+      productName
+      productKey
+      productSku
+      discountCode
+         
+####Array
+Define an Array property using names ending with "s".
```
Comment (jenschude) [inline API-GUIDELINE.md:229]:
```suggestion
Define an Array property using names in plural form, typically ending with "s".
```

There are mass nouns which have no singular form e.g. Money

File: API-GUIDELINE.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+Some examples:
+
+    discountCodes
+    refusedGifts
+    roles
+    allowedValues
+    
+####Object
+Define an Object property using "ResourceIdentifier" or "Draft" or "Reference" in the name.
```
Comment (jenschude) [inline API-GUIDELINE.md:239]: Below examples contradict this statement.

It's more that for objects there are not that much rules.

Draft is used for objects which are meant to be written

ResourceIdentifier for object which refer to other resource and are meant to be written

References for objects which refer to other resources and are only read

**Review by jenschude — CHANGES_REQUESTED** [review]

**Review by MicheleRezk — COMMENTED** [review]
LGTM :+1:

**Review by ashishhk — CHANGES_REQUESTED** [review]
I just added one comment so far,I'll add more once I read the rest of it.

File: API-GUIDELINE.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+####Boolean
+Define a Boolean property using a name which includes "is" or "has" and the verb at past participle.
+
+Some examples:
+
+        isEmailVerified
+        hasStagedChanges
+        isPublished
+        isActive
```
Comment (ashishhk) [inline API-GUIDELINE.md:204]: @jenschude you mentioned earlier sometime that we can recommend to write it without is or has as it might be redundant when we generate the SDKs.

e.g. active instead of isActive so that we generate a method named isActive based on the name.

WDYT ?

File: API-GUIDELINE.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+ DO'S
+ - use nouns instead of verbs to expose resources
+ - use lowercase and separated by hyphens (-) for path segments like /shopping-lists
+ - use American English
+ - create each resource as unique resource name
+ - use the proper HTTP methods for API operations whenever possible
+ 
+ DON'TS
+ - DON'T use snake case (_) or other punctuation
+ - DON'T use file extensions
```
Comment (ashishhk) [inline API-GUIDELINE.md:173]: possibly we can link this sentense to a set of global file extensions

**Review by ashishhk — COMMENTED** [review]
Added a few more comments.

File: API-GUIDELINE.md (hunk 5 above)
Comment (jenschude) [reply API-GUIDELINE.md:204]: It depends. On the pure API layer it's more speaking as the type can be read from the name.

For our SDKS its different. E.g. java says that a getter for a boolean bean should start with `is`. So we would have to declare mappings to avoid something like `isIsActive` or `getIsActive` (atm this is how it looks like in PHP)

And defacto the whole API already uses these prefixes and can't be easily changed

Comment (barbara79) [reply API-GUIDELINE.md:204]: so are you suggesting that I should change everything without "is"? Assuming that our guidelines is also for the future

Comment (barbara79) [reply API-GUIDELINE.md:204]: and what about something like "emailIsVerified"?

Comment (jenschude) [reply API-GUIDELINE.md:204]: if so it would be `emailVerified` -> singular rule ;)

And if we don't want to codify the state as is as best as possible, but make decisions about how it should be in the future, it should not only be us to do so ;)

Comment (barbara79) [reply API-GUIDELINE.md:204]: I agree but we should suggest something to the backend team, so I would go with `emailVerified`

Comment (jenschude) [reply API-GUIDELINE.md:204]: No we can't change existing fields. We could only propose how it should be done in the future.

Comment (barbara79) [reply API-GUIDELINE.md:204]: yes, this is what I was saying ;)

**Review by ashishhk — APPROVED** [review]
Looks very good to me, awesome work !!

File: API-GUIDELINE.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+Some examples:
+
+    shippingMethod
+    shippingRateInput
+    product
+    shippingDetailsToRemove
+    discountedPricePerQuantity
+    discountedPrice 
+
+####*Exceptions
```
[the remark this answers is not in the document]
Comment (barbara79) [reply API-GUIDELINE.md:252]: for my point of view yes, because it's a re-call of line 17 ;)

File: API-GUIDELINE.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
+#API GUIDELINES
```
Comment (yanns) [inline API-GUIDELINE.md:1]:
```suggestion
# API GUIDELINES
```
I think you need a space after all `#`, `##`...

File: API-GUIDELINE.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+##Overview
+As commercetools aims to be 100% API First, developer experience for API solution becomes very important for us.
```
Comment (yanns) [inline API-GUIDELINE.md:4]: > developer experience for API solution
I don't really understand this sentence. What do you mean by that?

File: API-GUIDELINE.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+ - Explicability => explain with a few words the purpose of the property avoiding abbreviations (there are some *exceptions).
+
+##Scope
+The purpose of these guidelines is to achieve the definition of consistent practices and patterns for all API endpoints.
+
+##URI
+Our URIs are composed by:
+   
+    Base Uri like:
+        https://api.{region}.{cloudProvider}.commercetools.com
```
Comment (yanns) [inline API-GUIDELINE.md:26]: `api` is one service.
We also have `auth` for example.

File: API-GUIDELINE.md (hunk 8 above)
Comment (xsoufiane) [reply API-GUIDELINE.md:1]: I saw that too 🤔 but after testing it worked on markdown intellij

**Review by yanns — COMMENTED** [review]
Thanks for doing this!

File: API-GUIDELINE.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (xsoufiane) [reply API-GUIDELINE.md:252]: this is so cool, I didn't know that you can reference a section in markdown too 🙏🏻 thank you. But I think you need to do \[exceptions\]\(#exceptions\) in line 17 to make the anchor work

File: API-GUIDELINE.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+ - GET -> retrieve data
+ - POST -> create or update data
+ - DELETE -> delete data
+
+It's important to avoid unexpected behavior. So for instance use GET to retrieve data and not to delete content!
+
+##Properties/Payload
+Clients interact with a service by exchanging representations of resources.
+
+An example, how it's right now the Cart payload:
```
Comment (xsoufiane) [inline API-GUIDELINE.md:59]: and maybe you can rephrase this to *An example of the Cart payload:* without the how and right now, because it seems like a question

**Review by xsoufiane — APPROVED** [review]
Great initiative 💯

File: API-GUIDELINE.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (barbara79) [reply API-GUIDELINE.md:252]: thanks for the info I'm gonna update it ASAP :)

File: API-GUIDELINE.md (hunk 8 above)
Comment (barbara79) [reply API-GUIDELINE.md:1]: I can confirm: in the Preview nothing change, every paragraph doesn't have the '#'

Comment (yanns) [reply API-GUIDELINE.md:1]: thx for checking.
I'would still prefer with a whitespace. It's easier to read the markdown.

File: API-GUIDELINE.md (hunk 9 above)
Comment (barbara79) [reply API-GUIDELINE.md:4]: I would rephrase saying: "As commercetools aims to be 100% API First, it's important to have a well designed developer experience"

Comment (yanns) [reply API-GUIDELINE.md:4]: "As commercetools is 100% API first, it's important to have a well designed developer experience"
😉

Comment (barbara79) [reply API-GUIDELINE.md:4]: 🥇

File: API-GUIDELINE.md (hunk 8 above)
Comment (barbara79) [reply API-GUIDELINE.md:1]: ok I'm gonna update for each of them

File: API-GUIDELINE.md (hunk 10 above)
Comment (barbara79) [reply API-GUIDELINE.md:26]: make sense I'm gonna specify that the URI sample is for the api service

Comment (barbara79) [reply API-GUIDELINE.md:26]: and for the other services there is the link below, what do you think about?

File: API-GUIDELINE.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+Resource is an object or representation of something, which has some associated data with it and can be helped by methods to operate on it.
+
+A resource can be an entity or a collection:
+ - Entity is a resource that has to be in singular.
+ - Collections are a set of resources that must be in plural.
+ 
+A resource may contain sub-collection resources.
+
+## Methods
+The HTTP verbs used are;
```
Comment (cneijenhuis) [inline API-GUIDELINE.md:48]:
```suggestion
The HTTP verbs used are:
```

File: API-GUIDELINE.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+ - GET -> retrieve data
+ - POST -> create or update data
```
Comment (cneijenhuis) [inline API-GUIDELINE.md:51]: IMO it's worth mentioning that a create should return a `201`, whereas an update should return a `200` HTTP status code.

**Review by cneijenhuis — COMMENTED** [review]
I think it is a great start, the things here look good to me!

However, I don't think it is complete. Consistency expands well beyond naming things.

I am not sure if this PR is supposed to be the start of the API Guidelines - in this case everything is fine and I'll approve it :)

Here are the topics I'd expect to see either in this PR or in follow up PRs:

- We follow standards where they are defined, e.g. HTTP + ecosystem, OAuth 2, GraphQL, CloudEvents
- JSON: When an optional value is not set, the field is missing (`null` is not used)
- Authentication (OAuth2, scopes naming)
- Error handling (https://docs.commercetools.com/http-api-errors, in particular https://docs.commercetools.com/http-api-errors#error-responses )
- Returning a list of resources, pagination: https://docs.commercetools.com/http-api#pagedqueryresult
- Default fields of a resource: `id`, `version`, `createdAt`, `lastModifiedAt` (and maybe https://docs.commercetools.com/http-api-client-logging - but that's in Beta)
- Semi-default field of a resource: `key`
- https://docs.commercetools.com/http-api#optimistic-concurrency-control

Also, GraphQL is completely missing. According to the Definition Of Done, it is required: https://wiki.commercetools.com/display/CTPROD/Definition+of+Done+-+Checklist

I am not sure on the following topics. They are standard for the Scala-backed APIs. I don't think other APIs necessarily need to provide these. However, it'd be bad/inconsistent if other APIs offer a same-ish feature in a different way:
- References and Reference Expansion
- Other common types like LocalizedString, Money, ...
- Query Language
- Correlation IDs
- Messages, Subscriptions, Extensions + architecturally a preference for event-driven systems over polling
- External OAuth
- Custom Types/Fields
- If CQRS, then https://docs.commercetools.com/http-api#partial-updates
(Maybe even more?)

Basically: If another API is having a data model or offering a feature in way X, your API should do it in the same way.

I know that for some features, it'd be best if the Scala services offers APIs that other services can use. We did that successfully with the Auth service: The importer has an own scope, and all services can use introspection to work with the tokens just like the Scala-backed APIs do.
For other things, the reality is unfortunately that another service won't be able to offer the feature, e.g. Messages and Subscriptions. But that can be changed 😉 

Sorry for the long review, overall it's a great start! Thanks for your work!!! 🏆

File: API-GUIDELINE.md (hunk 10 above)
Comment (yanns) [reply API-GUIDELINE.md:26]: which link? https://docs.commercetools.com/http-api.html#hosts ?
Have you seen https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server ?

File: API-GUIDELINE.md (hunk 13 above)
Comment (barbara79) [reply API-GUIDELINE.md:51]: it's just because there is people that use GET for everything ;)

File: API-GUIDELINE.md (hunk 10 above)
Comment (barbara79) [reply API-GUIDELINE.md:26]: yes and yes ;) but I can mention the link of the authorization if this is what you mean

Comment (yanns) [reply API-GUIDELINE.md:26]: My suggestion:
use `https://{service}.{region}.{cloudProvider}.commercetools.com `
and use the 2 links as examples.

**Review by yanns — APPROVED** [review]
good start!

File: API-GUIDELINE.md (hunk 13 above)
Comment (barbara79) [reply API-GUIDELINE.md:51]: I can fix it ;)

File: API-GUIDELINE.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+## Methods
+The HTTP verbs used are:
+
+ - GET -> retrieve data
+ - POST -> create data (it should return a 201 HTTP status code) or update data (it should return a 200 HTTP status code)
+ - DELETE -> delete data
+
+It's important to avoid unexpected behavior. So for instance use GET to retrieve data and not to delete content!
+
```
Comment (jenschude) [inline API-GUIDELINE.md:55]: Just realized QueryParameters are missing here too.

Some rules to think about:

- naming
- no change of return type based on parameters
- avoid XOR parameters

File: API-GUIDELINE.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    resources
+
+    by ID
+
+    by external ID with equals e.g. customers/key={key}
+
+See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server 
+
+## Resources
+Resource is an object or representation of something, which has some associated data with it and can be helped by methods to operate on it.
```
Comment (stmeissner) [inline API-GUIDELINE.md:39]: Do we want to be that generic or rather be more specific to our business domain?
```suggestion
A Resource is a data collection object or a data representation of an entity in the e-commerce domain, which is specified by data fields and methods to operate on them.
```

File: API-GUIDELINE.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+A resource can be an entity or a collection:
+ - Entity is a resource that has to be in singular.
+ - Collections are a set of resources that must be in plural.
+ 
+A resource may contain sub-collection resources.
+
+## Methods
+The HTTP verbs used are:
+
+ - GET -> retrieve data
```
Comment (stmeissner) [inline API-GUIDELINE.md:50]: Apart from GraphQL we always retrieve the full resource, right?
```suggestion
 - GET -> retrieve resource
```

File: API-GUIDELINE.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+ - POST -> create data (it should return a 201 HTTP status code) or update data (it should return a 200 HTTP status code)
```
Comment (stmeissner) [inline API-GUIDELINE.md:51]:
```suggestion
 - POST -> create resource (it should return a 201 HTTP status code) or update resource (it should return a 200 HTTP status code)
```

File: API-GUIDELINE.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+ - DELETE -> delete data
```
Comment (stmeissner) [inline API-GUIDELINE.md:52]:
```suggestion
 - DELETE -> delete resource
```

File: API-GUIDELINE.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+
+It's important to avoid unexpected behavior. So for instance use GET to retrieve data and not to delete content!
+
+## Properties/Payload
+Clients interact with a service by exchanging representations of resources.
```
Comment (stmeissner) [inline API-GUIDELINE.md:57]: Not only, also by providing URL parameters.

File: API-GUIDELINE.md [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+
+Some examples:
+
+        returnAt
+        expirationAt
+        validFrom
+        validTo
+
+#### Boolean
+Define a Boolean property using a name which NOT includes "is" but it can include "has" and the verb at past participle.
```
Comment (stmeissner) [inline API-GUIDELINE.md:197]:
```suggestion
Define a Boolean property by a the past participle form of a verb. 
Do NOT use the auxiliary verb "is" as prefix here, but "has" can be used in combination with a noun.
```

File: API-GUIDELINE.md [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+With this in mind, we have created "API Guidelines" that aims to help developers to design APIs:
+
+ - simple
+ - consistent 
+ - self-explanatory
+ 
+The Keywords to have a good Guidelines are:
+ 
+ - Consistency => following the Guidelines
+ - Uniformity => checking in the code if there are other properties with similar name and adjust in base of it or improve it
```
Comment (TohruKohrita) [inline API-GUIDELINE.md:16]: It is not clear what "adjust(ing) in base of it or improv(ing) it" means. "In base of" presumably means "on the basis of" or "based on"? What do you adjust? Perhaps adjust(ing) "field naming" based on it? The final phrase "or improve it" rather implies that there are other ways to improve the code, if not, maybe we can omit it. I suggest the following:

```suggestion
 - Uniformity => checking in the code if there are other properties with similar names and adjusting field naming based on it
```

File: API-GUIDELINE.md [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+     }
+
+## Naming convention
+
+### General Guidelines
+To use consistent names these are the rules to follow:
+
+ DO'S
+ - use nouns instead of verbs to expose resources
+ - use lowercase and separated by hyphens (-) for path segments like /shopping-lists
```
Comment (stmeissner) [inline API-GUIDELINE.md:166]: looks like we use lower camel case for property names composed of more than one word?

File: API-GUIDELINE.md [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+
+ - simple
+ - consistent 
+ - self-explanatory
+ 
+The Keywords to have a good Guidelines are:
+ 
+ - Consistency => following the Guidelines
+ - Uniformity => checking in the code if there are other properties with similar name and adjust in base of it or improve it
+ - Explicability => explain with a few words the purpose of the property avoiding abbreviations (there are some [exceptions](#exceptions)).
```
Comment (TohruKohrita) [inline API-GUIDELINE.md:17]:
```suggestion
 - Explicability => explaining in a few words the purpose of the property avoiding abbreviations (there are some [exceptions](#exceptions)).
```

Just because other items start with -ing.

File: API-GUIDELINE.md (hunk 16 above)
Comment (ashishhk) [reply API-GUIDELINE.md:50]: It's a good suggestion, we are using resource everywhere it will be very nice if the Guideline itself is consistent and following the convention.

File: API-GUIDELINE.md (hunk 22 above)
Comment (jenschude) [reply API-GUIDELINE.md:166]: In path segments not. For query parameters it's something different

File: API-GUIDELINE.md [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+
+## Scope
+The purpose of these guidelines is to achieve the definition of consistent practices and patterns for all API endpoints.
```
Comment (TohruKohrita) [inline API-GUIDELINE.md:20]:
```suggestion
The purpose of these guidelines is to define consistent practices and patterns for all API endpoints.
```

File: API-GUIDELINE.md (hunk 22 above)
Comment (stmeissner) [reply API-GUIDELINE.md:166]: sure, I mean, we need an additional bullet point about field names

File: API-GUIDELINE.md [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+      
+#### String
+Define a String property using names that have to be explicable about the content.
+
+Some examples:
+
+      orderEditResult
+      productName
+      productKey
+      productSku
```
Comment (stmeissner) [inline API-GUIDELINE.md:227]: so, in this case the acronym "SKU" will only keep the first letter capitalized?

File: API-GUIDELINE.md [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+    shippingMethod
+    shippingRateInput
+    product
+    shippingDetailsToRemove
+    discountedPricePerQuantity
+    discountedPrice 
+
+#### Exceptions
+The only exception allowed is:
+ - write "id" even if is an abbreviation
```
Comment (stmeissner) [inline API-GUIDELINE.md:254]: Let's add some paragraph about method names to this. Create methods are pretty straightforward, I think: "CreateResource"

For update actions we still hold onto this convention?
**set**Field: for optional properties, can be unset if value is empty
**change**Field: for mandatory fields, value cannot be empty

Retrieve methods:
**Query Resources** returns PagedQueryResult containing multiple resources (plural)
**Get Resource by Unique Identifier**, returns only the identified Resource (singular)

Unique Identifier can be UUID, Key, customer email, etc.

File: API-GUIDELINE.md [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+    by external ID with equals e.g. customers/key={key}
+
+See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server 
+
+## Resources
+A Resource is a data collection object or a data representation of an entity in the e-commerce domain, which is specified by properties and methods to operate on them.
+
+A resource can be an entity or a collection:
+ - Entity is a resource that has to be in singular.
+ - Collections are a set of resources that must be in plural.
```
Comment (TohruKohrita) [inline API-GUIDELINE.md:43]: Please ignore this comment if it is clear to developers (after all I think this is an internal document for technical people).

To me " ... are a set ..." sounds suspicious. This was the starting point.

I am not familiar with the technical things, so I am afraid of making unnecessary noise here, but still I do not understand lines 41-43. The lines 39 and 41 explains a resource in terms of an entity and a collection, where as the lines 42 and 43 explains an entity and collections in terms of resources. (eg. in l. 39, A Resource is ... or a data rep. of an entity; in l. 42, Entity is a resource ... )

Also, does the line 43 mean "A collection is a set of more than one resources"? If so, it contradicts the line 41, which in particular says that *a* resource can be a collection.

I am afraid that I simply missed some intended difference between **R**esource (l. 39) and **r**esource (else where) or I simply do not know the technical definition of the term "be in singular/plural". If this is the case, I am sorry.

Comment (barbara79) [reply API-GUIDELINE.md:43]: no problem Tohru, here the clarification in the line 39 there is the definition easily you can say that there is simple resource (it's just one result) = Entity or a collection resource(it's more then one result) = Collection.

File: API-GUIDELINE.md (hunk 25 above)
Comment (barbara79) [reply API-GUIDELINE.md:228]: yes, we are not used to have everything in capital letters as a property

File: API-GUIDELINE.md (hunk 27 above)
Comment (TohruKohrita) [reply API-GUIDELINE.md:43]: Thank you! Now I understand that the line 39 is the definition of a **Resource** in terms of data collection object and an entity. The line 39 itself is now clear. 

With this in mind, the line 41 seems to be either a rephrase of the line 39 or an explanation of the behavior of a single **R**esource. 

Either way, the lines 42 and 43 try to explain the words "entity" and "collection" used in the line 41, but it is done so in terms of Resources, which was defined in terms of entity and collection.
 
I am not sure but I am afraid that this is a circular argument or some word (collection, entity or R/resource) is used in more than one meaning.

So, I think there is a room for improvement here. But since I do not know what you mean to say, I can not unfortunately offer a suggestion. If the passage is clear to others, I am happy to let it go, but if not, maybe we can rephrase a bit?

**Review by nkuehn — COMMENTED** [review]
Hi, great initiative!  

This would deserve becoming more than just some markdown file once it's matured.   

- missing strong prescription not never leak internal values whose meaning cannot be entirely documented (for example that was the case with early lastModifiedBy  data)
- generally I suggest to use a more formal language in the document, using MUST / CAN /  SHOULD in a structured way as specced in https://tools.ietf.org/html/rfc2119 
- OAuth scopes naming convention missing
- convention for constant fixed values (--> e.g. for discriminators or predefined types / modes / API extension targets)
- I would love to see a definition list of reserved property names that have to have a certain meaning - esp. for those that are not domain properties but technical infrastructure (key, id, lastModifiedBy, version, ...).   Including a strict definition of the values (id: UUID string,  key: string,  version: integer number, .... )
- Describing our guidelines for flat vs. nested data structures would help a lot, too.   When do I use a nested object for some properties and when not.    An "array of" is easy.  Re-using existing data structures in some other place of the API is easy, too.  Using well defined public standards in a nested object is easy too (e.g. the GeoJSON Point in the channel).  But from then on it's often a personal opinion.  For example, in the Payments API "paymentMethodInfo" is a nested structure just for organizational purposes.  Is that actually improving the developer experience or not?  It's more organized when you see the JSON but I often saw devs not finding the properties that are nested and then misusing others 






PS: I have a history in formal requirements specification writing research, so take the feedback with that context ;-)

File: API-GUIDELINE.md (hunk 27 above)
Comment (barbara79) [reply API-GUIDELINE.md:43]: ok, I will think about it, thanks!

Comment (TohruKohrita) [reply API-GUIDELINE.md:43]: Just one final suggestion. If this is about the naming convention, you might want to write so more explicitly and perhaps with examples. 

An entity as a reference must be named in singular (eg. ... ).
A collection as a reference must be named in plural (eg. ... ).

**Comment by ashishhk** [comment]
> I think it is a great start, the things here look good to me!
> 
> However, I don't think it is complete. Consistency expands well beyond naming things.
> 
> I am not sure if this PR is supposed to be the start of the API Guidelines - in this case everything is fine and I'll approve it :)
> 
> Here are the topics I'd expect to see either in this PR or in follow up PRs:
> 
> * We follow standards where they are defined, e.g. HTTP + ecosystem, OAuth 2, GraphQL, CloudEvents
> * JSON: When an optional value is not set, the field is missing (`null` is not used)
> * Authentication (OAuth2, scopes naming)
> * Error handling (https://docs.commercetools.com/http-api-errors, in particular https://docs.commercetools.com/http-api-errors#error-responses )
> * Returning a list of resources, pagination: https://docs.commercetools.com/http-api#pagedqueryresult
> * Default fields of a resource: `id`, `version`, `createdAt`, `lastModifiedAt` (and maybe https://docs.commercetools.com/http-api-client-logging - but that's in Beta)
> * Semi-default field of a resource: `key`
> * https://docs.commercetools.com/http-api#optimistic-concurrency-control
> 
> Also, GraphQL is completely missing. According to the Definition Of Done, it is required: https://wiki.commercetools.com/display/CTPROD/Definition+of+Done+-+Checklist
> 
> I am not sure on the following topics. They are standard for the Scala-backed APIs. I don't think other APIs necessarily need to provide these. However, it'd be bad/inconsistent if other APIs offer a same-ish feature in a different way:
> 
> * References and Reference Expansion
> * Other common types like LocalizedString, Money, ...
> * Query Language
> * Correlation IDs
> * Messages, Subscriptions, Extensions + architecturally a preference for event-driven systems over polling
> * External OAuth
> * Custom Types/Fields
> * If CQRS, then https://docs.commercetools.com/http-api#partial-updates
>   (Maybe even more?)
> 
> Basically: If another API is having a data model or offering a feature in way X, your API should do it in the same way.
> 
> I know that for some features, it'd be best if the Scala services offers APIs that other services can use. We did that successfully with the Auth service: The importer has an own scope, and all services can use introspection to work with the tokens just like the Scala-backed APIs do.
> For other things, the reality is unfortunately that another service won't be able to offer the feature, e.g. Messages and Subscriptions. But that can be changed 😉
> 
> Sorry for the long review, overall it's a great start! Thanks for your work!!! 🏆

Thanks for the review and suggesting further topics that should be part of the Guidelines.
Yes, I agree that the Guidelines are not completed yet, and TBH, we were primarily focusing on the problems that Clients team faces, however we tried to gather feedback from multiple stakeholders and it wasn't as successful as the PR way of getting the feedback.
So we'll work on API Guidelines again together as a committee that I recently proposed and your suggestions should be considered as next scope of the guidelines.

cc: @barbara79

File: API-GUIDELINE.md [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+    resources
+
+    URI parameters like:
+        /{ID}
+        /customer-id={customerId}
+
+
+See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server
+
+### Query Parameters
```
Comment (barbara79) [inline API-GUIDELINE.md:39]: @jenschude could you please have a look on the query parameters paragraph?

File: API-GUIDELINE.md [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+The Query Parameters can be defined as the optional key-value and not unique that appear after the question mark(?) in the URL and they can be combined with other parameters with an ampersand(&).
```
Comment (jenschude) [inline API-GUIDELINE.md:40]: Should be rephrased as it's hard to understand

File: API-GUIDELINE.md [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+
+In this case the rules to follow are:
+ - follow the same naming convention as the properties
+ - no need to change of return type based on the parameters
```
Comment (jenschude) [inline API-GUIDELINE.md:44]:
```suggestion
 - don't change the response type based on query parameters
```

File: API-GUIDELINE.md (hunk 29 above)
Comment (barbara79) [reply API-GUIDELINE.md:40]: could you please check again? ;)

**Review by jenschude — APPROVED** [review]

File: API-GUIDELINE.md [hunk 31]
```diff
@@ -0,0 +1,282 @@
...
+    shippingRateInput
+    product
+    shippingDetailsToRemove
+    discountedPricePerQuantity
+    discountedPrice 
+
+#### Exceptions
+The only exception allowed is:
+ - write "Id" even if is an abbreviation
+ - write "Sku" even if is an abbreviation
```
Comment (lauraluiz) [inline API-GUIDELINE.md:282]:
```suggestion
 - write "Sku" even if is an acronym
```

**Review by lauraluiz — APPROVED** [review]
Thanks so much for starting this and pushing it forward! 🏆

**Comment by ashishhk** [comment]
> I am not sure on the following topics. They are standard for the Scala-backed APIs. I don't think other APIs necessarily need to provide these. However, it'd be bad/inconsistent if other APIs offer a same-ish feature in a different way:
> 
> * References and Reference Expansion
> * Other common types like LocalizedString, Money, ...
> * Query Language
> * Correlation IDs
> * Messages, Subscriptions, Extensions + architecturally a preference for event-driven systems over polling
> * External OAuth
> * Custom Types/Fields
> * If CQRS, then https://docs.commercetools.com/http-api#partial-updates
>   (Maybe even more?)
> 
> Basically: If another API is having a data model or offering a feature in way X, your API should do it in the same way.
> 
> I know that for some features, it'd be best if the Scala services offers APIs that other services can use. We did that successfully with the Auth service: The importer has an own scope, and all services can use introspection to work with the tokens just like the Scala-backed APIs do.
> For other things, the reality is unfortunately that another service won't be able to offer the feature, e.g. Messages and Subscriptions. But that can be changed 😉

Hi Christoph all your points should be taken care as part of these issues
https://github.com/commercetools/api-design-guidelines/issues/1
https://github.com/commercetools/api-design-guidelines/issues/2

File: API-GUIDELINE.md (hunk 25 above)
Comment (ashishhk) [reply API-GUIDELINE.md:254]: please follow further discussion here https://github.com/commercetools/api-design-guidelines/issues/3

File: API-GUIDELINE.md (hunk 26 above)
Comment (ashishhk) [reply API-GUIDELINE.md]: please follow further communication here https://github.com/commercetools/api-design-guidelines/issues/4

**Comment by ashishhk** [comment]
> Hi, great initiative!
> 
> This would deserve becoming more than just some markdown file once it's matured.
> 
> * missing strong prescription not never leak internal values whose meaning cannot be entirely documented (for example that was the case with early lastModifiedBy  data)
> * generally I suggest to use a more formal language in the document, using MUST / CAN /  SHOULD in a structured way as specced in https://tools.ietf.org/html/rfc2119
> * OAuth scopes naming convention missing
> * convention for constant fixed values (--> e.g. for discriminators or predefined types / modes / API extension targets)
> * I would love to see a definition list of reserved property names that have to have a certain meaning - esp. for those that are not domain properties but technical infrastructure (key, id, lastModifiedBy, version, ...).   Including a strict definition of the values (id: UUID string,  key: string,  version: integer number, .... )
> * Describing our guidelines for flat vs. nested data structures would help a lot, too.   When do I use a nested object for some properties and when not.    An "array of" is easy.  Re-using existing data structures in some other place of the API is easy, too.  Using well defined public standards in a nested object is easy too (e.g. the GeoJSON Point in the channel).  But from then on it's often a personal opinion.  For example, in the Payments API "paymentMethodInfo" is a nested structure just for organizational purposes.  Is that actually improving the developer experience or not?  It's more organized when you see the JSON but I often saw devs not finding the properties that are nested and then misusing others
> 
> PS: I have a history in formal requirements specification writing research, so take the feedback with that context ;-)

Thanks for the notes and help in PR, please follow here for the actions we'll take https://github.com/commercetools/api-design-guidelines/issues/5

**Comment by ashishhk** [comment]
PLEASE DO NOT DELETE THIS PR YET, however its closed for any further communications please follow any further communication here https://github.com/commercetools/api-design-guidelines
